dags/tmf/bikeshare/trips.py

```diff
@@ -0,0 +1,197 @@
+    # this DAG. Tableau server doesn't allow the download of underlying
+    # workbook data via the API (though one can from the UI). This view
+    # allows us to get around that.
+    logging.info("Loading Trips view")
+    all_views, _ = tableau_server.views.get()
+    view = next(v for v in all_views if v.id == TRIP_TABLE_VIEW_ID)
+    if not view:
+        raise Exception("Cannot find the trips table!")
+    tableau_server.views.populate_csv(view)
+    df = pandas.read_csv(
+        io.BytesIO(b"".join(view.csv)),
+        parse_dates=["Start Datetime", "End Datetime"],
+        thousands=",",
+        dtype={"visible_id": str, "end_station": str, "start_station": str},
+    )
+
+    # The data has a weird structure where trip rows are duplicated, with variations
+    # on a "Measure" column, containing trip length, duration, etc. We pivot on that
+    # column to create a normalized table containing one row per trip.
+    logging.info("Cleaning Data")
+    df = pandas.merge(
+        df.set_index("Trip ID")
+        .groupby(level=0)
+        .first()
+        .drop(columns=["Measure Names", "Measure Values"]),
+        df.pivot(index="Trip ID", columns="Measure Names", values="Measure Values"),
+        left_index=True,
+        right_index=True,
+    ).reset_index()
+    df = df.rename(
+        {
+            n: n.lower().strip().replace(" ", "_").replace("(", "").replace(")", "")
+            for n in df.columns
+        },
+        axis="columns",
+    )
+    check_columns(bike_trips, df)
+
+    # Upload the final dataframe to Postgres. Since pandas timestamps conform to the
+    # datetime interfaace, psycopg can correctly handle the timestamps upon insert.
+    logging.info("Uploading to PG")
+    engine = PostgresHook.get_hook(POSTGRES_ID).get_sqlalchemy_engine()
+    insert = sqlalchemy.dialects.postgresql.insert(bike_trips).on_conflict_do_nothing()
+    conn = engine.connect()
+    conn.execute(insert, *df.to_dict(orient="record"))
+
+
+def load_s3_data(**kwargs):
+    """
+    Load the table from PG and upload it as a parquet to S3.
+    """
+    bucket = kwargs.get("bucket")
+    name = kwargs.get("name", "bikeshare_trips.parquet")
+    if bucket:
+        logging.info("Uploading data to s3")
+        engine = PostgresHook.get_hook(POSTGRES_ID).get_sqlalchemy_engine()
+        df = pandas.read_sql_table(TABLE, engine, schema=SCHEMA)
+        path = os.path.join("/tmp", name)
+        df.to_parquet(path)
+        s3con = S3Hook(S3_ID)
+        s3con.load_file(path, name, bucket, replace=True)
+        os.remove(path)
+
+
+default_args = {
+    "owner": "ianrose",
+    "depends_on_past": False,
+    "start_date": datetime(2019, 11, 21),
+    "email": ["ian.rose@lacity.org"],
+    "email_on_failure": True,
+    "email_on_retry": False,
+    "retries": 1,
+    "retry_delay": timedelta(hours=1),
+}
+
+dag = DAG("bikeshare-trips", default_args=default_args, schedule_interval="@daily")
+
+t1 = PythonOperator(
```

Comment:
Thoughts on a python operator to create table or using a vanilla postgres operator to actually just create the table using `CREATE TABLE` ... 

The latter would allow us to use `CREATE IF DOES NOT EXIST`

Reply:
The SQLAlchemy table creation does `CREATE IF NOT EXIST` by default for tables: https://docs.sqlalchemy.org/en/13/core/tutorial.html#define-and-create-tables

It doesn't do that for schemas, unfortunately (probably an oversight on their part), hence  the somewhat awkward check here.

One of the reasons that I like assembling the SQLAlchemy table here (rather than raw SQL) is that I can use it below to check if the dataframe conforms to the schema before trying to insert.

Reply:
ah okay! LGTM